File: 160.intersection-of-two-linked-lists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def getIntersectionNode(self, headA, headB):
        """
        :type head1, head1: ListNode
        :rtype: ListNode
        """
        # Use two pointers and initialize them to headA/headB
        # when either is exhausted its list, redirect it to the other list
        # they either end up with same val node
        # or end up at the end of list at the same time on 2nd iteration
        if not headA or not headB: return None
        ptr1, ptr2 = headA, headB
        while ptr1 != ptr2:
            ptr1 = ptr1.next if ptr1 else headB
            ptr2 = ptr2.next if ptr2 else headA
        return ptr1

        

File: 160.intersection-of-two-linked-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        unordered_set<int> set;
        while(headA) {
            set.insert(headA->val);
            headA = headA->next;
        }
        while(headB) {
            if(set.find(headB->val) != set.end())
                return headB;
            headB = headB->next;
        }
        return NULL;
    }
};
